api/core/utils/redis_cache.py

```diff
@@ -57,3 +80,88 @@ async def async_wrapper(*args: Any, **kwargs: Any) -> Any:
+    """Helper function to cache a result."""
+    try:
+        await async_cache.setex(cache_key, expiration_seconds, pickle.dumps(result))  # pyright: ignore
+    except Exception as e:
+        _logger.exception("Failed to cache result for", exc_info=e, extra={"cache_key": cache_key})
+
+
+async def _safely_close_redis(async_cache: Any) -> None:
+    """Helper function to safely close Redis connections."""
+    if async_cache and hasattr(async_cache, "close") and callable(async_cache.close):
+        try:
+            await asyncio.shield(async_cache.close())
+        except Exception as e:
+            _logger.exception("Error while closing redis", exc_info=e)
+
+
+def redis_cached_generator_last_chunk(expiration_seconds: int = 60 * 60 * 24) -> Callable[[AG], AG]:
+    """
+    Decorator to cache the final chunk of an async generator function in Redis.
+
+    Limitations:
+        - Only the *final* result (the last yielded item) is cached.
+        - It assumes that the last item yielded by the generator represents the complete, cumulative result.
+        - It does not cache the intermediate streamed items.
+    """
+    async_cache = get_redis_client()
+    if not async_cache:
+        _logger.warning("Redis cache is not available, skipping redis_cached_generator")
+        return lambda func: func  # type: ignore
+
+    def decorator(func: AG) -> AG:
+        @functools.wraps(func)
+        async def async_generator_wrapper(*args: Any, **kwargs: Any) -> AsyncIterator[Any]:
+            cache_key = _generate_cache_key(func, args, kwargs, suffix=".generator_result")
+
+            try:
+                # Check for cached result
+                cached_item = await _try_retrieve_cached_result(async_cache, cache_key)
+                if cached_item is not None:
+                    # Cache hit - yield the cached item
+                    yield cached_item
+                    return
+
+                # Cache miss - run the generator
+                last_yielded = None
+
+                async for item in func(*args, **kwargs):
+                    last_yielded = item
+                    yield item
+
+                # Cache the last yielded item if there is one
+                if last_yielded is not None:
+                    await _try_cache_result(async_cache, cache_key, last_yielded, expiration_seconds)
+                else:
+                    _logger.warning(
+                        "Generator yielded no items for nothing to cache.",
+                        extra={"cache_key": cache_key},
+                    )
+
+            except Exception as e:
+                _logger.exception("Error in cached generator for", exc_info=e)
+                # Fallback to original function on error
+                async for item in func(*args, **kwargs):
+                    yield item
+            finally:
+                await _safely_close_redis(async_cache)
```

Comment:
does that work ?

You open the client once but close the connection after each call. Does redis automatically re-open connections when needed ? Seems like you should just be using a single, shared client ?

Reply:
shared client implemented here https://github.com/WorkflowAI/WorkflowAI/pull/34/commits/235c132ecf0f9d078e1b56441212ffa5f1158530

Reply:
Implement a shared cache in https://github.com/WorkflowAI/WorkflowAI/pull/34/commits/235c132ecf0f9d078e1b56441212ffa5f1158530